models/PyCryptoBot.py

```diff
@@ -204,19 +204,19 @@ def __init__(self, exchange='coinbasepro', filename='config.json'):
                                         self.simstartdate = config['simstartdate']
 
                         if 'sellupperpcnt' in config:
-                            if isinstance(config['sellupperpcnt'], int):
-                                if config['sellupperpcnt'] > 0 and config['sellupperpcnt'] <= 100:
-                                    self.sell_upper_pcnt = int(config['sellupperpcnt'])
+                            if isinstance(config['sellupperpcnt'], (int,str)):
```

Comment:
I don't think your validation will work here. If someone enters some junk as a string and it tries to cast as a float below it will crash. 

Please do something like this instead:
if isinstance(config['sellupperpcnt'], int) or isinstance(config['sellupperpcnt'], float):

Reply:
Good morning.

I initially tested the input for (int) and (float) but, a problem arises when you try to use (negative floats) in the json.

**A negative float is not accepted in the json format**. There are many cases of people referring to the issue (Google: "json and negative floats"). I did test using negative floats **without** quotes and the script (PyCryptoBot.py) crashed.

I didn't consider validating for junk in the string, thanks.

Below is a solution to check for the casting error so it exits out gracefully.

i.e.

```
try:
   if float(config['selllowerpcnt']) >= -100 and float(config['selllowerpcnt']) < 0:
         self.sell_lower_pcnt = float(config['selllowerpcnt'])
   except ValueError:
          print("Invalid format found. Use Integers or Float numbers")
          sys.exit()
```

I tested the code quickly with this:

```
#passed_float = "-.05"
passed_float = "pancakes"

try:
    if float(passed_float) >= -100 and float(passed_float) < 0:
        print("Success: value is within range.", "\n")
    else:
        print("Fail: float is not within range.", "\n")
except ValueError:
    print("Invalid format found. Use Integers or Float numbers ONLY")
    sys.exit()
```

If you're ok with that, I can make the changes and re-commit.

Reply:
That is interesting about the negative floats. If it needs to be an integer or string (for the float) you can use a regular expression to validate it.

    import re

    num = 10

    if isinstance(num, int):
        print ("OK")
    elif isinstance(num, str):
        p = re.compile(r"^[0-9\-\.]{1,3}$")
        if not p.match(num):
            print ("OK")
        else:
            print ("NOT OK")
    else:
        print ("NOT OK")

    num = -10

    if isinstance(num, int):
        print ("OK")
    elif isinstance(num, str):
        p = re.compile(r"^[0-9\-\.]{1,3}$")
        if not p.match(num):
            print ("OK")
        else:
            print ("NOT OK")
    else:
        print ("NOT OK")

    num = "1.5"

    if isinstance(num, int):
        print ("OK")
    elif isinstance(num, str):
        p = re.compile(r"^\-*[0-9\.]{1,5}$")
        if p.match(num):
            print ("OK")
        else:
            print ("NOT OK")
    else:
        print ("NOT OK")

    num = "-1.5"

    if isinstance(num, int):
        print ("OK")
    elif isinstance(num, str):
        p = re.compile(r"^\-*[0-9\.]{1,5}$")
        if p.match(num):
            print ("OK")
        else:
            print ("NOT OK")
    else:
        print ("NOT OK")

    num = "-.5"

    if isinstance(num, int):
        print ("OK")
    elif isinstance(num, str):
        p = re.compile(r"^\-*[0-9\.]{1,5}$")
        if p.match(num):
            print ("OK")
        else:
            print ("NOT OK")
    else:
        print ("NOT OK")

This prints:

    OK
    OK
    OK
    OK
    OK

Reply:
Hey @whittlem ,

I have pushed up the changes. I used the try/except variation coupled with just using the float function (**it performs regex internally to evaluate the passed values**). 

I also modified the command-line argument section to use (str) for the type values for (selllowerpcnt, sellupperpcnt, trailingstoploss).

I tested the logic in various scenarios via the command-line and from the configuration file.. And, I modified the error messages to be descriptive for both uses:

```
        if args.selllowerpcnt != None:
            if isinstance(args.selllowerpcnt, (int,str)):
                try:
                    if float(args.selllowerpcnt) >= -100 and float(args.selllowerpcnt) < 0:
                        self.sell_lower_pcnt = float(args.selllowerpcnt)
                except ValueError:
                    print("\nInvalid (selllowerpcnt) value in command-line arugment. Use Integers or Float numbers ONLY","\n")
                    sys.exit()
```

```
if 'selllowerpcnt' in config:
  if isinstance(config['selllowerpcnt'], (int,str)):
   try:
      if float(config['selllowerpcnt']) >= -100 and float(config['selllowerpcnt']) < 0:
          self.sell_lower_pcnt = float(config['selllowerpcnt'])
   except ValueError:
      print("\nInvalid (selllowerpcnt) value in configuration file. Use Integers or Float numbers ONLY","\n")
      sys.exit()
```